File: android/test-runner/src/main/kotlin/com/cavadvpn/vpn/VpnStats.kt
package com.cavadvpn.vpn

/**
 * Snapshot of VPN traffic statistics.
 *
 * @property bytesIn          Total bytes received from the server.
 * @property bytesOut         Total bytes sent to the server.
 * @property connectedSinceMs Wall-clock timestamp (ms) when the connection was established.
 * @property assignedIp       IP address assigned by the VPN server (e.g. "10.8.0.2").
 */
data class VpnStats(
    val bytesIn: Long,
    val bytesOut: Long,
    val connectedSinceMs: Long,
    val assignedIp: String
) {
    /** Elapsed time since connection, in milliseconds. */
    val uptimeMs: Long
        get() = if (connectedSinceMs <= 0L) 0L
                else System.currentTimeMillis() - connectedSinceMs

    /** Formatted uptime string "HH:MM:SS". */
    fun formatUptime(): String {
        val totalSeconds = (uptimeMs / 1000L).coerceAtLeast(0L)
        val hours   = totalSeconds / 3600
        val minutes = (totalSeconds % 3600) / 60
        val seconds = totalSeconds % 60
        return "%02d:%02d:%02d".format(hours, minutes, seconds)
    }

    companion object {
        /** An empty / zeroed-out stats object representing no active connection. */
        val EMPTY = VpnStats(
            bytesIn          = 0L,
            bytesOut         = 0L,
            connectedSinceMs = 0L,
            assignedIp       = ""
        )
    }
}

/**
 * Formats a byte count into a human-readable string with one decimal place.
 *
 * Examples:
 *   0         → "0 B"
 *   500       → "500 B"
 *   1500      → "1.5 KB"
 *   1_048_576 → "1.0 MB"
 *   1_073_741_824 → "1.0 GB"
 */
fun formatBytes(bytes: Long): String {
    return when {
        bytes < 1024L                -> "$bytes B"
        bytes < 1024L * 1024L       -> "%.1f KB".format(bytes / 1024.0)
        bytes < 1024L * 1024L * 1024L -> "%.1f MB".format(bytes / (1024.0 * 1024.0))
        else                        -> "%.1f GB".format(bytes / (1024.0 * 1024.0 * 1024.0))
    }
}

File: ios/App/Models/VpnStats.swift
// VpnStats.swift — Traffic statistics model for the iOS VPN UI

import Foundation

/// Traffic and connection statistics for a live VPN session.
public struct VpnStats: Equatable {
    /// Bytes received through the VPN tunnel.
    public var bytesIn: Int64
    /// Bytes sent through the VPN tunnel.
    public var bytesOut: Int64
    /// Unix timestamp (milliseconds) when the session started.
    public var connectedSinceMs: Int64
    /// IP address assigned to the client by the VPN server.
    public var assignedIP: String

    public static let empty = VpnStats(
        bytesIn: 0, bytesOut: 0,
        connectedSinceMs: Int64(Date().timeIntervalSince1970 * 1000),
        assignedIP: ""
    )

    public init(bytesIn: Int64, bytesOut: Int64, connectedSinceMs: Int64, assignedIP: String) {
        self.bytesIn = bytesIn
        self.bytesOut = bytesOut
        self.connectedSinceMs = connectedSinceMs
        self.assignedIP = assignedIP
    }

    /// Uptime in seconds.
    public var uptimeSeconds: Int64 {
        let nowMs = Int64(Date().timeIntervalSince1970 * 1000)
        return max(0, (nowMs - connectedSinceMs) / 1000)
    }

    /// Uptime formatted as HH:MM:SS.
    public var uptimeFormatted: String {
        let s = uptimeSeconds
        let h = s / 3600
        let m = (s % 3600) / 60
        let sec = s % 60
        return String(format: "%02d:%02d:%02d", h, m, sec)
    }

    /// Format a byte count as a human-readable string.
    public static func formatBytes(_ bytes: Int64) -> String {
        if bytes < 1024 {
            return "\(bytes) B"
        } else if bytes < 1024 * 1024 {
            return String(format: "%.1f KB", Double(bytes) / 1024)
        } else if bytes < 1024 * 1024 * 1024 {
            return String(format: "%.1f MB", Double(bytes) / (1024 * 1024))
        } else {
            return String(format: "%.1f GB", Double(bytes) / (1024 * 1024 * 1024))
        }
    }

    /// Formatted download label.
    public var downloadLabel: String { "↓ \(VpnStats.formatBytes(bytesIn))" }

    /// Formatted upload label.
    public var uploadLabel: String { "↑ \(VpnStats.formatBytes(bytesOut))" }
}
